modules/ximgproc/src/globalmatting.cpp

```diff
@@ -0,0 +1,616 @@
+        Mat_<uchar> &trimap,
+        int r, float c)
+{
+    int w = image.cols;
+    int h = image.rows;
+
+    for (int y = 0; y< w; ++y)
+    for (int x = 0; x< h; ++x)
+    {
+        if (trimap(y, x) != 128)
+            continue;
+
+        const Vec3b &I = image(y, x);
+
+        for (int j = y-r; j <= y+r; ++j)
+        {
+            for (int i = x-r; i <= x+r; ++i)
+            {
+                if (i < 0 || i >= w || j < 0 || j >= h)
+                    continue;
+
+                if (trimap(j, i) != 0 && trimap(j, i) != 255)
+                    continue;
+
+                const Vec3b &I2 = image(j, i);
+
+                float pd = sqrt((float)(sqr(x - i) + sqr(y - j)));
+                float cd = colorDist(I, I2);
+
+                if (pd <= r && cd <= c)
+                {
+                    if (trimap(j, i) == 0)
+                        trimap(y, x) = 1;
+                    else if (trimap(j, i) == 255)
+                        trimap(y, x) = 254;
+                }
+            }
+        }
+    }
+
+    for (int x = 0; x < trimap.cols; ++x)
+    {
+        for (int y = 0; y < trimap.rows; ++y)
+        {
+            if (trimap(y, x) == 1)
+                trimap(y, x) = 0;
+            else if (trimap(y, x) == 254)
+                trimap(y, x) = 255;
+        }
+    }
+}
+
+// erode foreground and background regions to increase the size of unknown region
+void GlobalMattingImpl::erodeFB(Mat_<uchar> &trimap, int r)
+{
+    int w = trimap.cols;
+    int h = trimap.rows;
+
+    Mat_<uchar> foreground(trimap.size(), (uchar)0);
+    Mat_<uchar> background(trimap.size(), (uchar)0);
+
+    for (int y = 0; y < h; ++y)
+    for (int x = 0; x < w; ++x)
+    {
+        if (trimap(y, x) == 0)
+            background(y, x) = 1;
+        else if (trimap(y, x) == 255)
+            foreground(y, x) = 1;
+    }
+
+
+    Mat kernel = getStructuringElement(MORPH_ELLIPSE, Size(r, r));
+
+    // FIXIT "Inplace" filtering call is ineffective in general (involves input data copying)
+    erode(background, background, kernel);
+    erode(foreground, foreground, kernel);
+
+    for (int y = 0; y < h; ++y)
+    {
+        for (int x = 0; x < w; ++x)
+        {
+            if (background(y, x) == 0 && foreground(y, x) == 0)
+                trimap(y, x) = 128;
+        }
+    }
+}
+
+
+void GlobalMattingImpl::calculateAlphaPatchMatch(const Mat_<Vec3b> &image,
+    const Mat_<uchar> &trimap,
+    const vector<Point> &foregroundBoundary,
+    const vector<Point> &backgroundBoundary,
+    vector<vector<Sample> > &samples)
+{
+    int w = image.cols;
+    int h = image.rows;
+
+    samples.resize(h, vector<Sample>(w));
+
+    for (int y = 0; y < h; ++y)
+    {
+        for (int x = 0; x < w; ++x)
+        {
+            if (trimap(y, x) == 128)
+            {
+                Point p(x, y);
+
+                samples[y][x].fi = rand() % foregroundBoundary.size();
+                samples[y][x].bj = rand() % backgroundBoundary.size();
```

Comment:
Avoid using `rand()` anywhere.
It is not stable and sometimes it is not thread safe. We need reproducible results for automated tests.

Consider using of OpenCV's thread local `theRNG()` implementation.